### PR TITLE
Prevent Qubit.__del__ from running more than once

### PR DESCRIPTION
**Pull request by Strilanc**
Currently, when the call to `engine.deallocate_qubit` fails, the `self.id = -1` line never runs. The failure can also resurrect the qubit object instance, so its del method gets scheduled to run again, and so an infinite error text loop closes.

Alternatively, the loop can be closed by "helpful" code triggered by the dealloc accidentally calling `__del__` on the qubit.

I've seen it happen (I injected a failure-on-tenth-iteration into LocalOptimizer than ran the Shor example). And this change fixed the problem.

### CODE REVIEW

File: projectq/types/_qubit.py [hunk 1]
```diff
@@ -116,12 +116,18 @@ class Qubit(BasicQubit):
     Thus the qubit is not copyable; only returns a reference to the same
     object.
     """
+    def __init__(self, engine, idx):
```
Comment (thomashaener) [inline projectq/types/_qubit.py]: I don't think we'll need __init__ anymore.

File: projectq/types/_qubit.py [hunk 2]
```diff
@@ -116,12 +116,18 @@ class Qubit(BasicQubit):
...
+        BasicQubit.__init__(self, engine, idx)
+
     def __del__(self):
         """
         Destroy the qubit and deallocate it (automatically).
         """
-        self.engine.deallocate_qubit(self)
+        if self.id == -1:
```
Comment (thomashaener) [inline projectq/types/_qubit.py]: `BasicEngine` handles this check already.

File: projectq/types/_qubit.py (hunk 1 above)
Comment (Strilanc) [inline projectq/types/_qubit.py]: Done

File: projectq/types/_qubit.py (hunk 2 above)
Comment (Strilanc) [inline projectq/types/_qubit.py]: There are already engines in the codebase with a `deallocate_qubit` method that don't do this check (in Simulator and in a mock test engine). This check reduces a complicated whole-code-base proof, about what's a main engine and what's a backend and how they get called, into a one-liner.

(... If it was up to me I'd make "you have to give me a valid qubit" a precondition of `engine.deallocate_qubit` and throw when it happened. Double-free's are symptoms of bugs.)

Comment (thomashaener) [inline projectq/types/_qubit.py]: Either is fine with me, as long as it's only in one location.

Comment (Strilanc) [inline projectq/types/_qubit.py]: Done. `deallocate_qubit` now asserts the qubit id isn't -1 as a precondition, and `__del__` satisfies that precondition by avoiding the call. People who have a bug that accidentally frees qubit twice will be appropriately warned by the exception.

Comment (thomashaener) [inline projectq/types/_qubit.py]: Thanks!

**Pull request merged**